### ISSUE TITLE
Plot completeness - purity

### ISSUE THREAD

**Issue by drphilmarshall** [drphilmarshall/PS1QLS#25]
Analogous to the `make_roc()` method, we need a method to plot _astronomical_ purity against completeness, assuming a "natural" abundance of let's say 1 lens per 1000 objects (although this will need to be a kwarg).

**Comment by jennykim1016** [comment]
I finished studying ROC curve in addition to reviewing some statistics. I think I figured out most of the things, but I wanted to check one thing before finishing the code.

Maybe I am just confused, but when drawing purity-completeness curve, how would the variable "abundance" matter? The pseudo code that we wrote on Tuesday is attached below.

`    def make_pc_plot(self, truth, abundance=1e-3):
        """
        Plots purity against completeness, assuming a natural abundance of lenses.
        
        Parameters
        ----------
        truth: ndarray
            True labels for the test dataset objects.
        abundance: float
            Fraction of all objects in the expected parent dataset that are actually lenses.`

I thought of using this method [http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html](url) to draw the purity-completeness curve. I think the purity-completeness curve and precision-recall curve are the same curves. However, because we receive "self" and "truth" arrays as our inputs, I believe we do not need to use "abundance" variable.

In other words, because the "truth" array knows what are the lenses and what are not, and because we are given "self" array, we could just draw the completeness-purity diagram with only those two arrays. If we would want to use "abundance" variable, I think we should not use one of the two arrays.

Would you mind clarifying what I am confused?

**Comment by drphilmarshall** [comment]
I'm glad you pointed out this resource: we should adopt the terms "precision" and "recall" when discussing a Classifier's performance on validation or test sets that have the same make-up as the training set. I read (and adapt) the following from the link you posted:
> The precision P is the ratio `TP / (TP + FP)` where `TP` is the number of true positives and `FP` the number of false positives. The precision is intuitively the ability of the classifier not to label as positive a sample that is negative.
The recall R is the ratio `TP / (TP + FN)` where `TP` is the number of true positives and `FN` the number of false negatives. The recall is intuitively the ability of the classifier to find all the positive samples.

When we give our classifier a test dataset that has *different* proportional class occupation, we need to account for that. Suppose we make a training set that is 50% lenses, 50% non-lenses, and we find that the precision is 90% and the recall is 75%. Now suppose that our astronomical test set contains 100,000 objects and only 100 of them are actually lenses. What do we expect our classifier to do?

Let's calculate the expected number of true positives, false negatives, true positives and true negatives in this case. Recall of 75% means that we expect the classifier to find 75 lenses, and miss 25, out of the 100 that are really there. That means that in the test set, `TP=75` and `FN=25` (it wrongly classified 25 non-lenses as lenses). 

Now we know what `TP` is, we could try plugging it into the classifier's precision formula and calculate an expected number of false positives, `FP`, as `FP = TP * (1-P)/P`. Since precision `P=0.9` we get `FP = 8.3`. However, this is the number of false positives we would expect *if the test set contained equal numbers of lenses and non-lenses*. This 8.3 is really "8.3 out of 100 non-lenses" and so is a false positive *rate*,`FPR = FP / (FP + TN)`. So let's multiply `FPR = 8.3 / 100` by the total number of non-lenses in our test set, 999,900 to get `FP = 82992`. The expected astronomical purity is then `AP = TP / (TP + FP) = 75 / (75 + 82992) = 0.09%` 

Hopefully this makes sense! Can you please derive a general formula for  expected astronomical purity in terms of TPR, FPR, precision, and recall given a 50-50 validation set, and an input natural lens abundance/What FPR are we aiming at if we want 10% purity?

**Comment by drphilmarshall** [comment]
BTW, note that Wikipedia is pretty good on statistics and machine learning. Here's a nice page with some good definitions on it that should help out with your manipulations:  https://en.wikipedia.org/wiki/Precision_and_recall

A graph of expected astronomical purity vs validation set FPR would be pretty useful, actually: it would help give us some intuition about these things. You could imagine plotting a few curves, for a variety of rare object abundance rates.

**Comment by drphilmarshall** [comment]
One last thing I forgot to write: the validation recall is the same as the expected astronomical completeness, in case that wasn't clear above.

**Comment by jennykim1016** [comment]
Thank you so much @drphilmarshall ! Your explanations and the Wikipedia article helped me a lot!

So we could get precision rate(let's use symbol "p" to denote the precision rate) and recall rate(let's use symbol "r" to denote the recall rate) by inputting the "self" and the "truth" array into this method [(http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html)](url). We know that recall rate "r" is **astronomical completeness**. 

Given a 50-50 validation set, true positive, false negative, false positive would be

> TP = 100 * r, FN = 100 * (1-r), FP = TP * ((1-p)/p)


By definition, we know that

> FPR = (FP) / (FP + TN)                 

> TPR = (TP) / (TP + FN)

where FP + TN = TP + FN = 100.

Because the **astronomical purity** is defined to be `FPR * (1-abundance)`, we could calculate astronomical purity. If we want to have one equation about **astronomical purity**, that would be `( (FP) / (FP + TN) ) * (1-abundance)`.

If we want astronomical purity to be 0.1, then `FPR` should be `0.1/(1-abundance)`.

So, I will plot **astronomical completeness** vs **astronomical purity** by following the calculations above.

**Comment by drphilmarshall** [comment]
Not quite: I think we'll need some different symbols, to distinguish the numbers in the astronomical test set from those in the training/validation set. Astronomical purity is `AP = ATP / (ATP + AFP)` where `ATP` is the number of true positives in the astronomical test set (75 in my example) and `AFP` is the number of false positives in the astronomical test set (8292 in my example). Let's keep "abundance" to mean the fraction of the astronomical test set that are actually lenses, 1e-3 in my example. `AP` is definitely related to the training set `FPR`, but not as simply as you have it, I think (try substituting numbers into your formula...)

**Comment by jennykim1016** [comment]
Oh, I think I was again confused. I will fix the problems and update it. Thanks for pointing it out!

**Comment by drphilmarshall** [comment]
No worries! If in doubt, revisit the notation and improve it until all is
clear...

**Comment by jennykim1016** [comment]
Again, we could get precision rate(let's use symbol "p" to denote the precision rate) and recall rate(let's use symbol "r" to denote the recall rate) by inputting the "self" and the "truth" array into this method [(http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html)](url). We know that recall rate "r" is **astronomical completeness**. In addition, r and p can range from 0 to 1.

Given a 50-50 validation set with  `S` elements, true positive, false negative, false positive would be

> TP = S * r, FN = S * (1-r), FP = TP * ((1-p)/p)

By definition, we know that

> FPR = (FP) / (FP + TN)                 

> TPR = (TP) / (TP + FN)

where `FP + TN = TP + FN = S`.

I think I was confused with the symbols this part. As you mentioned, the **astronomical purity** is defined to be `ATP / (ATP + AFP)`. Here,  `ATP = TPR * (Total number of Lenses)` and `AFP = FPR * S * (Total Number of Non-Lenses)`. If we call L as a number of the whole objects, the  `total number of lenses ` would be L * (abundance), and the  ` total number of non-lenses` would be L * (1-abundance). Thus, we can rewrite the equation to be  `ATP = TPR * L * abundance` and `AFP = FPR * S * L * (1-abundance).` If we plug these into the equation  `ATP / (ATP + AFP)`, we will get `Astronomical Purity = (TPR*L*abundance) / (TPR*L*abundance + FPR*S*L*(1-abundance))`. The constant  `L` cancels out, so this can be simplified as : ` Astronomical Purity = (TPR*abundance) / (TPR*abundance + FPR*(1-abundance)) ` .

If I plug in S = 100, L = 100,000 to the final equation, I get the exactly the same values as you calculated above. 

So, I will plot **astronomical completeness** vs **astronomical purity** by following the calculations above.

**Comment by drphilmarshall** [comment]
Very nice. Looking fwd to seeing the plot in the notebook!
